src/store_cache.rs

```diff
@@ -0,0 +1,78 @@
+use std::cmp::max;
+use std::io::Read;
+use super::hyper::client::Client as HttpClient;
+use super::hyper::header::{
+    CacheControl as HyCacheControl,
+    CacheDirective as HyCacheDirective
+};
+use super::redis::{Commands};
+use super::error::BrokerResult;
+use super::store::Store;
+
+
+/// Represents a Redis key.
+pub enum CacheKey<'a> {
+    Discovery { domain: &'a str },
+    KeySet { domain: &'a str },
+}
+
+impl<'a> CacheKey<'a> {
+    fn to_string(&self) -> String {
+        match *self {
+            CacheKey::Discovery { domain } => {
+                format!("cache:discovery:{}", domain)
+            },
+            CacheKey::KeySet { domain } => {
+                format!("cache:key-set:{}", domain)
+            }
+        }
+    }
+}
+
+
+#[derive(Clone, Deserialize)]
+pub struct StoreCache {
+    pub expire_keys: usize, // Redis key minimum TTL, in seconds
+}
+
+impl StoreCache {
+    /// Fetch `url` from cache or using a HTTP GET request. The cache is stored in `store` with
+    /// `key`. The `session` is used to make the HTTP GET request, if necessary.
+    pub fn fetch_url(&self, store: &Store, key: CacheKey, session: &HttpClient, url: &str)
+                     -> BrokerResult<String> {
+
+        // Format the Redis key.
```

Comment:
IMO the comments in this function are a bit too fine-grained/are largely repeating the code (that is, they don't add that much value). I think they could be removed and/or changed to provide more rationale/why rather than what is happening.